x/epochs/types/cache_ctx.go

```diff
@@ -0,0 +1,83 @@
+package types
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime"
+	"runtime/debug"
+
+	"cosmossdk.io/core/appmodule"
+	storetypes "cosmossdk.io/store/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// This function lets you run the function f, but if there's an error or panic
+// drop the state machine change and log the error.
+// If there is no error, proceeds as normal (but with some slowdown due to SDK store weirdness)
+// Try to avoid usage of iterators in f.
+//
+// If its an out of gas panic, this function will also panic like in normal tx execution flow.
+// This is still safe for beginblock / endblock code though, as they do not have out of gas panics.
+func ApplyFuncIfNoError(ctx context.Context, env appmodule.Environment, f func(ctx context.Context) error) (err error) {
```

Comment:
Can we delete this and just use the branch service at the place it is needed?

Reply:
I made a refactor for hook, no need panic recover anymore